### PR TITLE
Feat(lockdrop): ability to claim bluna rewards

### PR DESCRIPTION
**Pull request by epanchee**

### CODE REVIEW

File: contracts/lockdrop/src/contract.rs [hunk 1]
```diff
@@ -255,8 +284,52 @@ pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> StdResult<Binary> {
...
 
 #[cfg_attr(not(feature = "library"), entry_point)]
-pub fn migrate(_deps: DepsMut, _env: Env, _msg: MigrateMsg) -> StdResult<Response> {
-    Ok(Response::default())
+pub fn migrate(deps: DepsMut, _env: Env, _msg: MigrateMsg) -> StdResult<Response> {
+    use std::str;
+
+    let contract_version = get_contract_version(deps.storage)?;
+
+    let response = Response::new()
```
[the remark this answers is not in the document]
Comment (epanchee) [reply contracts/lockdrop/src/contract.rs:292]: I think it does not make sense

File: contracts/lockdrop/src/contract.rs [hunk 2]
```diff
@@ -1126,6 +1200,77 @@ pub fn handle_claim_rewards_and_unlock_for_lockup(
     Ok(Response::new().add_messages(cosmos_msgs))
 }
 
+fn handle_claim_asset_reward(
```
[the remark this answers is not in the document]
Comment (epanchee) [reply contracts/lockdrop/src/contract.rs:1206]: done

File: contracts/lockdrop/src/contract.rs [hunk 3]
```diff
@@ -255,8 +284,52 @@ pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> StdResult<Binary> {
...
+                    };
+                    ASSET_POOLS.save(deps.storage, &key, &new_pool_info)?
+                }
+            }
+            _ => return Err(StdError::generic_err("Migration error")),
+        },
+        _ => return Err(StdError::generic_err("Migration error")),
+    };
+
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
```
Comment (P-Yevhenii) [inline contracts/lockdrop/src/contract.rs:330]: be cool to add CONTRACT_NAME and CONTRACT_VERSION to the response.

Comment (epanchee) [reply contracts/lockdrop/src/contract.rs:326]: Ok

File: contracts/lockdrop/src/contract.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (epanchee) [reply contracts/lockdrop/src/contract.rs]: Done. Thanks

File: contracts/lockdrop/src/contract.rs [hunk 4]
```diff
@@ -1848,6 +2056,37 @@ fn calculate_weight(amount: Uint128, duration: u64, config: &Config) -> Uint256
...
+    }
+
+    let to_distribute = match USERS_BLUNA_REWARD_INDEX.may_load(deps.storage, user)? {
+        None => total_reward_balance,
+        Some(last_user_bluna_reward_index)
+            if last_user_bluna_reward_index < total_reward_balance =>
+        {
+            total_reward_balance - last_user_bluna_reward_index
+        }
+        _ => Uint256::zero(),
```
Comment (P-Yevhenii) [inline contracts/lockdrop/src/contract.rs:2079]: Here, we can return Uint256:zero() and skip steps below.

File: contracts/lockdrop/src/contract.rs [hunk 5]
```diff
@@ -1848,6 +2056,37 @@ fn calculate_weight(amount: Uint128, duration: u64, config: &Config) -> Uint256
...
+    };
+
+    USERS_BLUNA_REWARD_INDEX.save(deps.storage, user, &total_reward_balance)?;
```
Comment (P-Yevhenii) [inline contracts/lockdrop/src/contract.rs:2082]: We can move save user index to 2077 line and save gas.

Comment (epanchee) [inline contracts/lockdrop/src/contract.rs:2082]: I think your suggestion above will cover this

File: contracts/lockdrop/src/contract.rs (hunk 4 above)
Comment (epanchee) [reply contracts/lockdrop/src/contract.rs]: Sure, thank you.

File: contracts/lockdrop/src/contract.rs [hunk 6]
```diff
@@ -1513,6 +1664,63 @@ pub fn callback_deposit_liquidity_in_astroport(
...
+    mut deps: DepsMut,
+    env: Env,
+    previous_balance: Uint128,
+    terraswap_lp_token: Addr,
+    user_address: Addr,
+    lock_duration: u64,
+) -> StdResult<Response> {
+    let reward_balance =
+        astroport::querier::query_balance(&deps.querier, env.contract.address, "uusd".to_string())?;
+    let latest_reward_amount = reward_balance - previous_balance;
```
Comment (P-Yevhenii) [inline contracts/lockdrop/src/contract.rs:1677]: If we did not receive rewards, then latest_reward_amount will be zero. We should add a check for this one, and don't save the state if total_bluna_reward  has not changed.

File: contracts/lockdrop/src/contract.rs [hunk 7]
```diff
@@ -1513,6 +1664,70 @@ pub fn callback_deposit_liquidity_in_astroport(
...
+        .add_attribute("user", user_address.clone());
+
+    let total_bluna_reward = match TOTAL_ASSET_REWARD.may_load(deps.storage)? {
+        Some(balance) if !latest_reward_amount.is_zero() => {
+            let val = balance + Uint256::from(latest_reward_amount);
+            TOTAL_ASSET_REWARD.save(deps.storage, &val)?;
+            val
+        }
+        None => {
+            let val = reward_balance.into();
```
[the remark this answers is not in the document]
Comment (epanchee) [reply contracts/lockdrop/src/contract.rs]: Ok, fixed.

File: contracts/lockdrop/src/contract.rs [hunk 8]
```diff
@@ -1126,6 +1201,82 @@ pub fn handle_claim_rewards_and_unlock_for_lockup(
...
 
+/// ## Description
+/// Collects assets reward from LP and distribute reward to user if all requirements are met.
+/// Otherwise returns [`StdError`].
+fn handle_claim_asset_reward(
+    deps: Deps,
+    env: Env,
+    user_address: Addr,
+    terraswap_lp_token: String,
+    lock_duration: u64,
```
[the remark this answers is not in the document]
Comment (epanchee) [reply contracts/lockdrop/src/contract.rs:1212]: Ok, updated code in `callback_distribute_asset_reward` (I perform this check in this func)

File: contracts/lockdrop/src/state.rs [hunk 9]
```diff
@@ -14,6 +14,10 @@ pub const ASSET_POOLS: Map<&Addr, PoolInfo> = Map::new("LiquidityPools");
 pub const USER_INFO: Map<&Addr, UserInfo> = Map::new("users");
 /// Key consists of an Terraswap LP token address, an user address, and a duration
 pub const LOCKUP_INFO: Map<(&Addr, &Addr, U64Key), LockupInfo> = Map::new("lockup_position");
+/// Total received asset reward by lockdrop contract
+pub const TOTAL_ASSET_REWARD: Item<Uint256> = Item::new("total_asset_reward");
+/// Latest used total asset reward for user claim
+pub const USERS_ASSET_REWARD_INDEX: Map<&Addr, Uint256> = Map::new("users_asset_reward_index");
```
[the remark this answers is not in the document]
Comment (epanchee) [reply contracts/lockdrop/src/state.rs]: Ok. updated these constants

File: contracts/lockdrop/src/contract.rs (hunk 6 above)
Comment (epanchee) [reply contracts/lockdrop/src/contract.rs:1677]: I fixed this case

File: contracts/lockdrop/src/contract.rs [hunk 10]
```diff
@@ -1513,6 +1664,70 @@ pub fn callback_deposit_liquidity_in_astroport(
...
+        &terraswap_lp_token,
+        &user_address,
+        U64Key::new(lock_duration),
+    );
+    let mut user_reward = Uint128::zero();
+    let lockup_info_opt = LOCKUP_INFO.may_load(deps.storage, lockup_key.clone())?;
+    if let Some(lockup_info) = lockup_info_opt {
+        let pool_info = ASSET_POOLS.load(deps.storage, &terraswap_lp_token)?;
+
+        user_reward = calc_user_reward(
```
[the remark this answers is not in the document]
Comment (epanchee) [reply contracts/lockdrop/src/contract.rs:1738]: Added this functionality

File: contracts/lockdrop/src/state.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (ShadoySV) [reply contracts/lockdrop/src/state.rs]: (lp_token, user_addr) key can be more practical for iterating

[the remark this answers is not in the document]
Comment (epanchee) [reply contracts/lockdrop/src/state.rs]: Ok, fixed

File: contracts/lockdrop/src/contract.rs [hunk 11]
```diff
@@ -147,6 +149,27 @@ pub fn execute(deps: DepsMut, env: Env, info: MessageInfo, msg: ExecuteMsg) -> S
...
+            duration,
+        } => {
+            let user_address = recipient.map_or_else(
+                || Ok(info.sender),
+                |recip_addr| addr_validate_to_lower(deps.api, &recip_addr),
+            )?;
+            handle_claim_asset_reward(
+                deps.as_ref(),
+                env,
+                user_address,
```
[the remark this answers is not in the document]
Comment (epanchee) [reply contracts/lockdrop/src/contract.rs]: Fixed

[the remark this answers is not in the document]
Comment (epanchee) [reply contracts/lockdrop/src/contract.rs]: Fixed

**Review by ShadoySV — APPROVED** [review]

**Pull request merged**